File: python/rearrangeLinkedList.py
class LinkedList:
    def __init__(self, value):
        self.value = value
        self.next = None
        
        

# Time -> 0(N) Space -> 0(1)
def rearrangeLinkedList(head: LinkedList, k: int):
    newHead, less, startMerge, endMerge, startGreat, great = None, None, None, None, None, None
    
    curr = head
    while curr != None:
        if curr.value < k:
            if newHead == None:
                newHead = curr
                less = newHead
            else:
                less.next = curr
                less = less.next
        elif curr.value > k:
            if startGreat == None:
                startGreat = curr
                great = startGreat
            else:
                great.next = curr
                great = great.next
        else:
            if startMerge == None:
                startMerge = curr
                endMerge = startMerge
            else:
                endMerge.next = curr
                endMerge = endMerge.next
                
        curr = curr.next
        
    if newHead == None or startGreat == None or startMerge == None:
        if newHead == None:
            if startMerge == None:
                return startGreat
            else:
                if startGreat == None:
                    return startMerge
                else:
                    endMerge.next = startGreat
                    great.next = None
                    return startMerge
            
        else:
            if startMerge == None:
                if startGreat == None:
                    return newHead
                else:
                    less.next = startGreat
                    great.next = None
                    return newHead
            else:
                less.next = startMerge
                endMerge.next = None
                return newHead

                    
    less.next = startMerge
    endMerge.next = startGreat
    great.next = None
    
    return newHead

head = LinkedList(0)
curr = head
array = [3, 0, 5, 2, 1, 4]

for val in range(len(array)):
    curr.value = array[val]
    if val == len(array) - 1:
        break
    curr.next = LinkedList(0)
    curr = curr.next
    
arrange = rearrangeLinkedList(head, 3)

while arrange != None:
    print(arrange.value, end="->")
    arrange = arrange.next
print()

File: go/rearrangeLinkedList/rearrangeLinkedList.go
package main

import "fmt"


type LinkedList struct {
	Value int
	Next *LinkedList
}


// Time -> 0(N) Space -> 0(1)
func RearrangeLinkedList(head *LinkedList, k int) *LinkedList{
	var new_head, less, startGreat, great, startMerge, endMerge *LinkedList 

	curr := head
	for curr != nil {
		if curr.Value < k {
			if less == nil {
				less = curr
				new_head = less
			}else {
				less.Next = curr
				less = less.Next
			}
		}else if curr.Value > k {	
			if great == nil {
				great = curr
				startGreat = great
			}else {
				great.Next = curr
				great = great.Next
			}
		}else {
			if startMerge == nil {
				startMerge = curr
				endMerge = startMerge
			} else {
				endMerge.Next = curr
				endMerge = endMerge.Next
			}
		}

		curr = curr.Next
	}

	if new_head == nil || startMerge == nil || startGreat == nil {
		if new_head == nil {
			if startMerge == nil {
				return startGreat
			}else {
				if startGreat == nil {
					return startMerge
				}else {
					endMerge.Next = startGreat
					great.Next = nil
					return startMerge
				}
			}
		}else {
			if startMerge == nil {
				if startGreat == nil {
					return new_head
				}else {
					less.Next = startGreat
					great.Next = nil
					return new_head
				}
			}else {
				less.Next = startMerge
				endMerge.Next = nil
				return new_head
			}
		}


	}

	less.Next = startMerge
	endMerge.Next = startGreat
	great.Next = nil

	return new_head

}


func main() {

	list := []int{7, 8, 9, 10, 0}
	head := &LinkedList{}
	curr := head
	for i, val := range list {
		curr.Value = val
		if i == len(list) - 1 {continue}
		curr.Next = &LinkedList{}
		curr = curr.Next
	}

	rearranged := RearrangeLinkedList(head, 3)
	for rearranged != nil {
		fmt.Printf("%v -> ", rearranged.Value)
		rearranged = rearranged.Next
	}
	fmt.Println()
}